### ISSUE TITLE
[req]: Add Chinese language support

### ISSUE THREAD

**Issue by fangguisheng** [Jackett/Jackett#13262]
### Is there already a request for your feature?

- [X] I have checked older issues, open and closed

### Is your feature request related to a problem? Please describe.

Hope you can add language options. I can help complete the Chinese language

### Describe the solution you'd like

Many users in China are using Jackett, which is very good, but shy of novices. It is not friendly enough. I try to change the XML file directly into Chinese, but I will not package it into a docker image, so I hope to increase the language switching. I will complete the Chinese work

**Comment by ilike2burnthing** [comment]
As well as the main UI, translations would have to be done for the over 500 indexers, any future changes to those indexers (which happens regularly), and all new indexers (for example, we added 2 today).

This would be a massive and ongoing job, and I don't really see it as being fair or practical to expect one person to do it (and really, with any translation work for any project you need at least two people, so that work can be checked).

For those reasons, I think we'd have to decline your offer. But I'll get a second opinion from @garfield69 before closing this.

**Comment by fangguisheng** [comment]
> 除了主 UI 之外，还必须为 500 多个索引器、将来对这些索引器的任何更改（定期发生）以及所有新的索引器（例如，我们今天添加了 2 个索引器）进行翻译。
> 
> 这将是一项庞大而持续的工作，我真的不认为期望一个人来做这件事是公平或实际的（实际上，对于任何项目的任何翻译工作，你至少需要两个人，以便可以检查工作）。
> 
> 出于这些原因，我认为我们必须拒绝您的报价。但是在结束之前，我会从@garfield69那里得到第二个意见。

Hello, actually, you don't need to translate the content in the indexer. All PT players have a certain foundation. You only need to translate the UI part. I also have a communication group of various PT sites, and we will work together to test

**Comment by fangguisheng** [comment]
> 除了主 UI 之外，还必须为 500 多个索引器、将来对这些索引器的任何更改（定期发生）以及所有新的索引器（例如，我们今天添加了 2 个索引器）进行翻译。
> 
> 这将是一项庞大而持续的工作，我真的不认为期望一个人来做这件事是公平或实际的（实际上，对于任何项目的任何翻译工作，你至少需要两个人，以便可以检查工作）。
> 
> 出于这些原因，我认为我们必须拒绝您的报价。但是在结束之前，我会从@garfield69那里得到第二个意见。

Hello, actually, you don't need to translate the content in the indexer. All PT players have a certain foundation. You only need to translate the UI part. I also have a communication group of various PT sites, and we will work together to test

**Comment by ilike2burnthing** [comment]
Many indexers include instructions for how to configure the tracker, account settings, how to grab cookies, notes to use FlareSolverr, the sorting and order in which results are requested from the tracker, etc.. Presumably if there is want to have the UI translated, these would need translated as well.

**Comment by fangguisheng** [comment]
> Many indexers include instructions for how to configure the tracker, account settings, how to grab cookies, notes to use FlareSolverr, the sorting and order in which results are requested from the tracker, etc.. Presumably if there is want to have the UI translated, these would need translated as well.

There is no problem with this. The UI translation is not difficult and the content is basically fixed. I lack a switch to switch languages and call Chinese.xml interface

**Comment by ilike2burnthing** [comment]
I'm confused why there wouldn't be a problem in reading the English instructions and settings for each indexer, but there would be a problem in reading the English main UI.

**Comment by fangguisheng** [comment]
> 我很困惑为什么阅读每个索引器的英语说明和设置不会有问题，但是在阅读英语主UI时会有问题。

The indexer settings only need to be translated to most common interfaces for the configuration of URLs, accounts, passwords, cookies and so on. As a user who uses Jackett for the first time, he will feel friendly at first

**Comment by ilike2burnthing** [comment]
Here are a few examples of more complex indexer instructions:

> <ol><li>Login to this tracker with your browser<li>Open the <b>DevTools</b> panel by pressing <b>F12</b><li>Select the <b>Network</b> tab<li>Click on the <b>Doc</b> button (Chrome Browser) or <b>HTML</b> button (FireFox)<li>Refresh the page by pressing <b>F5</b><li>Click on the first row entry<li>Select the <b>Headers</b> tab on the Right panel<li>Find <b>'cookie:'</b> in the <b>Request Headers</b> section<li><b>Select</b> and <b>Copy</b> the whole cookie string <i>(everything after 'cookie: ')</i> and <b>Paste</b> here.</ol>

> This site may use Cloudflare DDoS Protection, therefore Jackett requires <a href="https://github.com/Jackett/Jackett#configuring-flaresolverr" target="_blank">FlareSolverr</a> to access it.

> As the iTorrents .torrent download link on this site is known to fail from time to time, we suggest using the magnet link as a fallback. The BTCache and Torrage services are not supported because they require additional user interaction (a captcha for BTCache and a download button on Torrage.)

> For best results, change the <b>Torrents per page:</b> setting to <b>100</b> on your account profile.

> Find or Generate a new API Token by accessing your <a href="https://web.site/" target =_blank>Website here</a> account <i>My Security</i> page and clicking on the <b>API Token</b> tab.

> Find or Generate a new RSS key by accessing your <a href="https://web.site/" target =_blank>Website here</a> account <i>My Security</i> page and clicking on the <b>RSS Key (RID)</b> tab.

There are also some less common options like:

> Replace MULTI by another language in release name

>Replace MULTI by this language

> Replace VOSTFR with ENGLISH

If someone can read and follow these, surely they don't need the main UI translated.

**Comment by fangguisheng** [comment]
You are right. I can use it in English now, but when I first use it, I need to use translation tools + search engines to determine what I really need to do, including the steps of grabbing cookies and torrents per page. For users who first contact Jackett, their mother tongue will be very useful. I still hope to win the following for Chinese users. After all, it is only a UI interface, which is not very difficult, thanks

**Comment by garfield69** [comment]
The Jackett WEB GUI is not a static presentation of HTML. The content is dynamically generated using input from C# modules.
Users that select a language will expect all of the page to be in their chosen language, not just the GUI headers and buttons labels.
This means that all the C# indexers and the Cardigann yaml processor (which generates dynamically changing text based on the yaml indexer requirements), contribute to the text that gets presented on the WEB GUI page, so these C# modules will need amending to support using a translation service.
My first point here is that this will require contribution by a C# developer, which for this project are in very short supply, with irregular attendance and short contribution windows.

Once the translation support code is implemented, then there are the tables that contain the language key/value pairs that will need to be created, and updated regularly to cater for changes that crop up in the lifetime of an active indexer.
Translation files for the most popular languages will need to be created, which, based on the indexers that we currently support, should include Chinese, Russian, French, Spanish, German, Italian, Korean, Japanese, Dutch, Czech, Greek, Hebrew, Hungarian, Polish, Turkish, Vietnamese, and possible some others.
My second point here is that this is no trivial task, and will require many hours of work and ongoing commitment to translation table maintenance.

A simpler solution for users that require translation support, is for them to use a Browser that has the built in Google translation service, such as Chrome or Edge, which has settings to automatically translate a page into the local language of the user. Its not perfect but better than nothing.

Alternatively, if anyone want to submit a proof of concept PR that provides multi language support to Jackett, then that would be a start.

**Comment by dxbzth** [comment]
Want to add Chinese support

**Comment by ngosang** [comment]
This is not a priority for me but it's doable. IMHO:

* We should only translate the WEB UI. The Web UI changes few times and it has few texts, so, the translations will be easy to maintain.
* This could be done without touching a single line of C#. If we wan''t to save the chosen language in the Jackett config, we have to add 1 configuration option in C# but it's not mandatory. We can save the language in the LocalStorage of the browser (as we do with the filters).
* The source code must remain in English language. It will be 1 JSON file for each language. and that will be the only place when "non-english" language will be allowed. This JS library is able to download the "es.json    fr.json.." files and translate the text in the client site (browser) => For example, using this library we can have => https://github.com/wikimedia/jquery.i18n
* I think it's a really bad idea to translate the "indexer configurations" or include translated text in the Cardigann definitions. I talking about the custom configuration specific of one indexer, most common configuration fields like URL, User, Password... could be translated with the rest of the WebUI.
* The only exception to the precious point could be localized indexers. For example if some indexer has the Web UI and the results in Chinese, we could assume it will be used only by Chinese speakers. In that case the indexer could be DUAL, English / Chinese. But this is not related to the Web UI translation, we re just including English and 1 other language "strings" in the definition to make things clear.
* We are not talking about the documentation but it will be important too. We should translate the readme and the troubleshooting guide at least. We have the wiki for that.

**Comment by mikeoscar2006** [comment]
Few points I had in mind:
1. We can use an automated translation service for most of the strings.
2. When a user changes the language, It can take a few minutes to cache all the translations and then change the whole UI.
3. Each following translation request can be cached as well.
4. Users can report whether there is a major & consistent issue in a translation.
5. For this, we can keep language override files.
6. This way, we'll only need to maintain such inconsistent translations.

Just my 2 cents on the issue. Although I +1 @garfield69's idea to keep things simple which is use the browser in built Translate Page option which we already use if we use a non-English tracker.

**Comment by Altansar69** [comment]
if it ever comes to fruition, I could help out as a French translator if needed.
